Fix warnings when compiling stdlib with --test

fhahn opened the pull request:
This PR siliences some warnings when compiling stdlib with --test. Mostly remove some unused imports and added a few `#[allow(..)]`.

I also marked some signal handling functions with `#[cfg(not(test))]`, because they are only called through `rt::lang_start`, which is also marked as  `#[cfg(not(test))]`

rust-highfive commented:
r? @aturon

(rust_highfive has picked a reviewer for you, use r? to override)

steveklabnik commented:
Looks good to me, but this isn't my specialty in the codebase, so I'll let someone else r+

alexcrichton commented on src/libstd/sys/unix/mod.rs:
```diff
@@ -59,6 +60,8 @@ pub fn init() {
 }
+
+#[cfg(not(test))]
 #[cfg(target_os = "nacl")]
```
I think these `cfg` annotations should be combined, right?

alexcrichton commented on src/libstd/sys/unix/stack_overflow.rs:
```diff
@@ -55,8 +57,10 @@ mod imp {
 
 
     // This is initialized in init() and only read from after
+    #[cfg(not(test))]
```
Hm instead of sprinkling these around, perhaps this could be at the top with something like `#![cfg_attr(test, allow(dead_code))]`?

alexcrichton commented on src/libstd/sys/common/remutex.rs:
```diff
@@ -205,6 +204,7 @@ mod tests {
     }
 
     #[test]
+    #[allow(unused_variables)]
```
Could the unused variables be removed here?

fhahn replied:
I think removing the unused variables here would change the test, because the locks are stored in variables to extend their lifetime to the end of the block, right?

alexcrichton replied:
You can just rename to `let _foo = expr` in this test to avoid adding the attribute

alexcrichton commented on src/libstd/sys/common/remutex.rs:
```diff
@@ -225,6 +225,8 @@ mod tests {
 
     #[test]
+    #[allow(unreachable_code)]
+    #[allow(unused_variables)]
```
Perhaps the unused variables and unreachable code could be removed?

alexcrichton commented on src/libstd/sys/common/io.rs:
```diff
@@ -129,6 +129,7 @@ mod tests {
     }
 
     #[bench]
+    #[allow(unused_must_use)]
```
Could this return the `Result` by just passing it through as the return value of the closure?

fhahn commented:
@alexcrichton Thanks for your feedback! I've updated the commit. Now almost all warnings on Linux should be gone. The only remaining warnings are about the use of deprecated items in src/libstd/dynamic_lib.rs (https://github.com/rust-lang/rust/pull/30458/files#diff-59794c31034a5097419f3f376a30b4a4R135 and https://github.com/rust-lang/rust/pull/30458/files#diff-59794c31034a5097419f3f376a30b4a4R168), even though I've added `#[allow(deprecated)]`.

alexcrichton commented on src/libstd/sys/common/remutex.rs:
```diff
@@ -233,9 +233,8 @@ mod tests {
             let lock2 = mc.lock().unwrap();
             *lock.borrow_mut() = 2;
-            let answer = Answer(lock2);
+            Answer(lock2);
```
I think this may have slightly changed the semantics here as `Answer` lives over the call to `panic!`,  but changing the declaration to `let _answer = ...` and removing the `drop` should have the same semantics.

alexcrichton commented:
Thanks @fhahn!  Looks good to me!

fhahn commented:
@alexcrichton thanks for your feedback! I've updated the PR again.

alexcrichton commented:
@bors: r+ abe9b4088989912f0cfa2348e6a103131791f3df

Thanks!

bors commented:
:hourglass: Testing commit abe9b40 with merge efa0caf...

bors commented:
:broken_heart: Test failed - [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/7472)

fhahn commented:
I've update the PR to fix the fail on android

sanxiyn commented:
@bors r+

bors commented:
:pushpin: Commit 217bebb has been approved by `sanxiyn`

<!-- @bors r=sanxiyn 217bebbda20e549140476d2adcce6decb1af9681 -->

bors commented:
:hourglass: Testing commit 217bebb with merge 1d71545...

bors commented:
:broken_heart: Test failed - [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/7498)

fhahn commented:
The build on Android fails, because I marked [`os::raw::tests::unix`](https://github.com/rust-lang/rust/pull/30458/files#diff-c4f9e907b21efdcec1a2c091ff9ee641R92) as test. I could disable it again for Android (or in general), but shouldn"t the types of `nlink_t` in `libc` and `os::raw` be equal?

sanxiyn commented:
I agree they should match on Android too, but I think it could be done on a separate PR.

fhahn commented:
I've excluded the test from Android for now (at least I hope) and opened a PR for [libc](https://github.com/rust-lang-nursery/libc/pull/119). After this lands, we should enable the test for android again.

sanxiyn commented:
@bors r+

bors commented:
:pushpin: Commit e27cbef has been approved by `sanxiyn`

<!-- @bors r=sanxiyn e27cbeff370897b8450caa204c08049651a10c13 -->

bors commented:
:hourglass: Testing commit e27cbef with merge a06bb97...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/7547), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/7565), [auto-linux-64-debug-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-debug-opt/builds/821), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/7502), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/7523), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/7506), [auto-linux-cross-opt](http://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/731), [auto-linux-musl-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-musl-64-opt/builds/2684), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/7575), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/7567), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/7533), [auto-win-gnu-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-nopt-t/builds/2536), [auto-win-gnu-32-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/2550), [auto-win-gnu-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-nopt-t/builds/2553), [auto-win-gnu-64-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/2533), [auto-win-msvc-32-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/1785), [auto-win-msvc-64-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/2533)
